src/components/ipDetector.ts

```diff
@@ -0,0 +1,122 @@
+import IDetector from './../interfaces/IDetector';
+import IIpDeviceManager from '../interfaces/iIpDeviceManager';
+import DetectorOpts from './../interfaces/IDetectorOpts';
+import CameraEvents from './../utilitis/events';
+import DetectionsConverter from './../utilitis/detectionsConverter';
+import * as huddly from '@huddly/camera-proto/lib/api/huddly_pb';
+import Logger from './../utilitis/logger';
+import { Empty } from 'google-protobuf/google/protobuf/empty_pb';
+
+const PREVIEW_IMAGE_SIZE = { width: 832, height: 480 };
+
+export default class IpDetector extends EventEmitter implements IDetector {
+  _deviceManager: IIpDeviceManager;
+  _detectionHandler: any;
+  _framingHandler: any;
+  _frame: any;
+  _options: DetectorOpts;
+  _intervalId: NodeJS.Timeout;
+  _defaultLabelWhiteList: Array<String> = ['head', 'person'];
+  _subscriptionsSetup: boolean = false;
+  _detectorInitialized: boolean = false;
+  _previewStreamStarted: boolean = false;
+  _usePeopleCount: boolean = false;
+  _UPDATE_INTERVAL: number = 500; // ms
+  _detectorWasStarted: boolean = false;
+
+  constructor(manager: IIpDeviceManager, options?: DetectorOpts) {
+    super();
+    this._deviceManager = manager;
+    this._options = options || {}; // validateOptions ?
+    this._options.DOWS = (options && options.DOWS) || false;
+  }
+
+  async init(): Promise<any> {
+    if (this._detectorInitialized) {
+      Logger.warn('Detector already initialized', 'L1 Detector');
+      return;
+    }
+    Logger.debug('Initializing detector class', 'L1 Detector');
+
+    // _detectionHandler fetches detections from the grpc server and emits them.
+    if (!this._detectionHandler) {
+      this._detectionHandler = async () => {
+        const detections = await this._getDetections();
+        const convertedDetections = this.convertDetections(detections, this._options);
+        this.emit(CameraEvents.DETECTIONS, convertedDetections);
+      };
+    }
+    if (!this._options.DOWS) {
+      Logger.debug('Sending start-cnn Detector command', 'L1 Detector');
+      await this._startDetections();
+    }
+    this._intervalId = setInterval(this._detectionHandler, this._UPDATE_INTERVAL);
+    this._detectorInitialized = true;
+    Logger.debug('Detector class initialized and ready', 'L1 Detector');
+  }
+
+  _getDetections(): Promise<Array<any>> {
+    return new Promise((resolve, reject) => {
+      this._deviceManager.grpcClient.getDetections(
+        new Empty(),
+        (err, detections: huddly.Detections) => {
+          if (err != undefined) {
+            Logger.error('Unable to get detections', err.stack || '');
+            reject(err.message || 'Unknown error');
+          }
+          resolve(detections.toObject().detectionsList || []);
+        }
+      );
+    });
+  }
+
+  _startDetections(): Promise<huddly.DeviceStatus> {
+    if (this._detectorWasStarted) {
+      Logger.error('Detector already started', '', 'L1 Detector');
+      return Promise.reject();
```

Comment:
Reject or Resolve? Maybe you dont want to reject promise if they call the start method more than once?

Reply:
This is just a way to avoid successive grpc start/stop calls. If you really want to directly send the grpc start/stop messages, you can do this through Ace.ts and not these private methods. If the implementation of the IpDetector class is correct, this clause should never happen anyway, so I can change it if you think it's better that way.